File: packages/rspack/src/loader-runner/contentFilter.ts
/**
 * Content-based loader filter for the JS side of the loader runner.
 *
 * One Rust->JS crossing carries an entire contiguous run of JS
 * loaders: the Normal loop below walks the loader list itself and only
 * yields back to Rust at a `builtin:` loader. The native filter in
 * `rspack_binding_api` therefore only ever sees the head of a run; for
 * every loader below the head the decision has to be taken here, or
 * not at all.
 *
 * Reads the same `RSPACK_LOADER_CONTENT_FILTER` JSON as the native
 * side: `[{"loader": "<request substring>", "include": ["needle"],
 * "includeRegex": ["pattern"]}]`. A gated loader whose current content
 * matches no needle and no pattern is skipped as identity: content,
 * source map and additional data pass through unchanged. Buffers are
 * searched bytewise for literals, without decoding.
 *
 * `loader` also matches against `<request> id=<options.id>`, which
 * distinguishes registrations that share one loader file (rsbuild's
 * `api.transform` gives every transform the same `transformLoader.mjs`
 * request and tells them apart only by its options). A real
 * implementation attaches the filter to the rule use item and needs no
 * such key.
 */

type ContentGate = {
  loader: string;
  include: string[];
  includeBytes: Buffer[];
  includeRegex: RegExp[];
};

let cachedGates: ContentGate[] | null = null;

function gates(): ContentGate[] {
  if (cachedGates !== null) return cachedGates;
  cachedGates = [];
  const raw = process.env.RSPACK_LOADER_CONTENT_FILTER;
  if (!raw) return cachedGates;
  try {
    const parsed = JSON.parse(raw);
    if (Array.isArray(parsed)) {
      for (const item of parsed) {
        const loader = typeof item?.loader === 'string' ? item.loader : '';
        const include = Array.isArray(item?.include)
          ? item.include.filter((m: unknown) => typeof m === 'string')
          : [];
        const includeRegex: RegExp[] = [];
        if (Array.isArray(item?.includeRegex)) {
          for (const pattern of item.includeRegex) {
            if (typeof pattern !== 'string') continue;
            try {
              includeRegex.push(new RegExp(pattern));
            } catch (e) {
              console.error(
                `[rspack] invalid content filter regex ${pattern}: ${e}`,
              );
            }
          }
        }
        if (loader && (include.length || includeRegex.length)) {
          cachedGates.push({
            loader,
            include,
            includeBytes: include.map((m: string) => Buffer.from(m)),
            includeRegex,
          });
        }
      }
    }
  } catch (e) {
    console.error(`[rspack] invalid RSPACK_LOADER_CONTENT_FILTER: ${e}`);
  }
  return cachedGates;
}

let skipped = 0;
let ran = 0;
let statsHookInstalled = false;

function countSkip(didSkip: boolean) {
  if (process.env.RSPACK_LOADER_CONTENT_FILTER_STATS !== '1') return;
  if (didSkip) skipped++;
  else ran++;
  if (!statsHookInstalled) {
    statsHookInstalled = true;
    process.on('exit', () => {
      console.error(
        `[rspack] content-filter(js): skipped=${skipped} ran=${ran}`,
      );
    });
  }
}

/**
 * True when this loader is gated and `content` cannot match, so the
 * loader's normal function can be skipped as identity.
 */
export function contentFilterShouldSkip(
  request: string,
  content: unknown,
  options?: unknown,
): boolean {
  const all = gates();
  if (all.length === 0) return false;
  const id =
    options && typeof options === 'object' && 'id' in options
      ? (options as { id?: unknown }).id
      : undefined;
  const key = typeof id === 'string' ? `${request} id=${id}` : request;
  const gate = all.find((g) => key.includes(g.loader));
  if (!gate) return false;
  let matched: boolean;
  if (typeof content === 'string') {
    matched =
      gate.include.some((m) => content.includes(m)) ||
      gate.includeRegex.some((re) => re.test(content));
  } else if (Buffer.isBuffer(content)) {
    matched = gate.includeBytes.some((m) => content.includes(m));
    if (!matched && gate.includeRegex.length) {
      const text = content.toString('utf8');
      matched = gate.includeRegex.some((re) => re.test(text));
    }
  } else {
    // Unknown content shape: never skip.
    return false;
  }
  countSkip(!matched);
  return !matched;
}

File: crates/rspack_binding_api/src/plugins/js_loader/content_filter.rs
//! Content-based loader filter, evaluated before yielding to JS.
//!
//! A JS loader that can only act on modules containing a marker symbol
//! still costs one Rust->JS crossing for every module it is matched
//! against, because rule conditions are evaluated at factory time,
//! before content is read. `loader_should_yield` fires after content is
//! read and before the yield, so a content test here removes the
//! crossing outright for every non-matching module.
//!
//! Skipping is identity: declining the yield falls through to the
//! loader trait's no-op default (`JsLoader` never overrides `run()`),
//! so content, source map and additional data pass through unchanged.
//!
//! Configuration comes from `RSPACK_LOADER_CONTENT_FILTER`, a JSON
//! array. `loader` is matched as a substring of the loader request; a
//! module whose current content matches none of `include` (literal
//! needles) or `includeRegex` (regular expressions) is skipped for
//! that loader:
//!
//! ```json
//! [{"loader": "transform.mjs", "include": ["createServerFn"],
//!   "includeRegex": ["\\.\\s*handler\\s*\\("]}]
//! ```
//!
//! The same variable is read by the JS loader runner in
//! `@rspack/core`, which walks contiguous runs of JS loaders without
//! returning to Rust: a filter evaluated only here is unreachable for
//! every loader below the head of a run.
use std::sync::{
  OnceLock,
  atomic::{AtomicU64, Ordering},
};

use rspack_loader_runner::Content;

pub static CONTENT_FILTER_SKIPPED: AtomicU64 = AtomicU64::new(0);
pub static CONTENT_FILTER_CROSSED: AtomicU64 = AtomicU64::new(0);

#[derive(Debug)]
struct ContentGate {
  loader: String,
  include: Vec<String>,
  include_regex: Vec<regex::bytes::Regex>,
}

fn gates() -> &'static Vec<ContentGate> {
  static GATES: OnceLock<Vec<ContentGate>> = OnceLock::new();
  GATES.get_or_init(|| {
    let Ok(raw) = std::env::var("RSPACK_LOADER_CONTENT_FILTER") else {
      return Vec::new();
    };
    let parsed: serde_json::Value = match serde_json::from_str(&raw) {
      Ok(v) => v,
      Err(e) => {
        eprintln!("[rspack] invalid RSPACK_LOADER_CONTENT_FILTER: {e}");
        return Vec::new();
      }
    };
    let mut out = Vec::new();
    if let Some(items) = parsed.as_array() {
      for item in items {
        let loader = item
          .get("loader")
          .and_then(|v| v.as_str())
          .unwrap_or_default()
          .to_string();
        let include: Vec<String> = item
          .get("include")
          .and_then(|v| v.as_array())
          .map(|a| {
            a.iter()
              .filter_map(|m| m.as_str().map(|s| s.to_string()))
              .collect()
          })
          .unwrap_or_default();
        let mut include_regex = Vec::new();
        if let Some(list) = item.get("includeRegex").and_then(|v| v.as_array()) {
          for pattern in list.iter().filter_map(|p| p.as_str()) {
            match regex::bytes::Regex::new(pattern) {
              Ok(re) => include_regex.push(re),
              Err(e) => eprintln!("[rspack] invalid content filter regex {pattern:?}: {e}"),
            }
          }
        }
        if !loader.is_empty() && (!include.is_empty() || !include_regex.is_empty()) {
          out.push(ContentGate {
            loader,
            include,
            include_regex,
          });
        }
      }
    }
    out
  })
}

fn stats_enabled() -> bool {
  static ENABLED: OnceLock<bool> = OnceLock::new();
  *ENABLED.get_or_init(|| std::env::var("RSPACK_LOADER_CONTENT_FILTER_STATS").as_deref() == Ok("1"))
}

/// True when this loader is gated and the module content cannot match.
pub fn should_skip(loader_request: &str, content: Option<&Content>) -> bool {
  let gates = gates();
  if gates.is_empty() {
    return false;
  }
  let Some(gate) = gates.iter().find(|g| loader_request.contains(&g.loader)) else {
    return false;
  };
  let Some(content) = content else {
    return false;
  };
  // Marker search over raw bytes: never decodes, never allocates.
  let bytes = content.as_bytes();
  let matched = gate
    .include
    .iter()
    .any(|m| memchr::memmem::find(bytes, m.as_bytes()).is_some())
    || gate.include_regex.iter().any(|re| re.is_match(bytes));
  if matched {
    CONTENT_FILTER_CROSSED.fetch_add(1, Ordering::Relaxed);
    false
  } else {
    let n = CONTENT_FILTER_SKIPPED.fetch_add(1, Ordering::Relaxed) + 1;
    if stats_enabled() && n % 10_000 == 0 {
      eprintln!(
        "[rspack] content-filter(rust): skipped={} crossed={}",
        n,
        CONTENT_FILTER_CROSSED.load(Ordering::Relaxed)
      );
    }
    true
  }
}
